<launch>
  <!-- Right arm -->
  <group ns="r_gripper_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_gripper_joint" />
    <param name="pid/p" type="double" value="100.0" />
    <param name="pid/i" type="double" value="1" />
    <param name="pid/d" type="double" value="0.0" />
    <param name="pid/i_clamp" type="double" value="0.5" />
  </group>
    
  <group ns="r_wrist_roll_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_wrist_roll_joint" />
    <param name="pid/p" type="double" value="12.0" />
    <param name="pid/i" type="double" value="3" />
    <param name="pid/d" type="double" value="0.5" />
    <param name="pid/i_clamp" type="double" value="1" />
  </group>
  
  <group ns="r_wrist_flex_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_wrist_flex_joint" />
    <param name="pid/p" type="double" value="12" />
    <param name="pid/i" type="double" value="3" />
    <param name="pid/d" type="double" value="0.5" />
    <param name="pid/i_clamp" type="double" value="1" />
  </group>

  <group ns="r_forearm_roll_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_forearm_roll_joint" />
    <param name="pid/p" type="double" value="20.0" />
    <param name="pid/i" type="double" value="5" />
    <param name="pid/d" type="double" value="0.5" />
    <param name="pid/i_clamp" type="double" value="2" />
  </group>
    
  <group ns="r_elbow_flex_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_elbow_flex_joint" />
    <param name="pid/p" type="double" value="35.0" />
    <param name="pid/i" type="double" value="5" />
    <param name="pid/d" type="double" value="1.0" />
    <param name="pid/i_clamp" type="double" value="2" />
  </group>
  
  <group ns="r_upper_arm_roll_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_upper_arm_roll_joint" />
    <param name="pid/p" type="double" value="20" />
    <param name="pid/i" type="double" value="2" />
    <param name="pid/d" type="double" value="2.0" />
    <param name="pid/i_clamp" type="double" value="1.0" />
  </group>

  <group ns="r_shoulder_lift_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_shoulder_lift_joint" />
    <param name="pid/p" type="double" value="40.0" />
    <param name="pid/i" type="double" value="7" />
    <param name="pid/d" type="double" value="2.0" />
    <param name="pid/i_clamp" type="double" value="4" />
  </group>
  
  <group ns="r_shoulder_pan_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_shoulder_pan_joint" />
    <param name="pid/p" type="double" value="50" />
    <param name="pid/i" type="double" value="10" />
    <param name="pid/d" type="double" value="5.0" />
    <param name="pid/i_clamp" type="double" value="3.0" />
  </group>

  <!-- Left arm -->
  <group ns="l_gripper_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_gripper_joint" />
    <param name="pid/p" type="double" value="100.0" />
    <param name="pid/i" type="double" value="1" />
    <param name="pid/d" type="double" value="0.0" />
    <param name="pid/i_clamp" type="double" value="0.5" />
  </group>
  
  
  <group ns="l_wrist_roll_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_wrist_roll_joint" />
    <param name="pid/p" type="double" value="12.0" />
    <param name="pid/i" type="double" value="3" />
    <param name="pid/d" type="double" value="0.5" />
    <param name="pid/i_clamp" type="double" value="1" />
  </group>
  
  <group ns="l_wrist_flex_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_wrist_flex_joint" />
    <param name="pid/p" type="double" value="12" />
    <param name="pid/i" type="double" value="3" />
    <param name="pid/d" type="double" value="0.5" />
    <param name="pid/i_clamp" type="double" value="1" />
  </group>

  <group ns="l_forearm_roll_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_forearm_roll_joint" />
    <param name="pid/p" type="double" value="20.0" />
    <param name="pid/i" type="double" value="5" />
    <param name="pid/d" type="double" value="0.5" />
    <param name="pid/i_clamp" type="double" value="2" />
  </group>
    
  <group ns="l_elbow_flex_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_elbow_flex_joint" />
    <param name="pid/p" type="double" value="35.0" />
    <param name="pid/i" type="double" value="5" />
    <param name="pid/d" type="double" value="1.0" />
    <param name="pid/i_clamp" type="double" value="2" />
  </group>
  
  <group ns="l_upper_arm_roll_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_upper_arm_roll_joint" />
    <param name="pid/p" type="double" value="20" />
    <param name="pid/i" type="double" value="2" />
    <param name="pid/d" type="double" value="2.0" />
    <param name="pid/i_clamp" type="double" value="1.0" />
  </group>

  <group ns="l_shoulder_lift_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_shoulder_lift_joint" />
    <param name="pid/p" type="double" value="40.0" />
    <param name="pid/i" type="double" value="7" />
    <param name="pid/d" type="double" value="2.0" />
    <param name="pid/i_clamp" type="double" value="4" />
  </group>
  
  <group ns="l_shoulder_pan_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="r_shoulder_pan_joint" />
    <param name="pid/p" type="double" value="50" />
    <param name="pid/i" type="double" value="10" />
    <param name="pid/d" type="double" value="5.0" />
    <param name="pid/i_clamp" type="double" value="3.0" />
  </group>

  <!-- Torso -->
  <group ns="torso_lift_joint_hold" clear_params="true">
    <param name="type" type="string" value="JointPositionController" />
    <param name="joint" type="string" value="torso_lift_joint" />
    <param name="pid/p" type="double" value="2000000" />
    <param name="pid/i" type="double" value="1000" />
    <param name="pid/d" type="double" value="0.0" />
    <param name="pid/i_clamp" type="double" value="1200.0" />
  </group>
  
  <node pkg="pr2_mechanism_control" type="spawner.py" args="r_gripper_joint_hold r_wrist_roll_joint_hold r_wrist_flex_joint_hold r_forearm_roll_joint_hold r_elbow_flex_joint_hold r_upper_arm_roll_joint_hold r_shoulder_lift_joint_hold r_shoulder_pan_joint_hold l_gripper_joint_hold l_wrist_roll_joint_hold l_wrist_flex_joint_hold l_forearm_roll_joint_hold l_elbow_flex_joint_hold l_upper_arm_roll_joint_hold l_shoulder_lift_joint_hold l_shoulder_pan_joint_hold" />

</launch>
